docs/rules/no-side-effects-wait-for.md

```diff
@@ -0,0 +1,71 @@
+# Side effects inside `waitFor` are not preferred (no-side-effects-wait-for)
+
+## Rule Details
+
+This rule aims to avoid the usage of side effects actions (`fireEvent` or `userEvent`) inside `waitFor`.
+Since `waitFor` is intended for things that have a non-deterministic amount of time between the action you performed and the assertion passing,
+the callback can be called (or checked for errors) a non-deterministic number of times and frequency.
+This will make your side-effect run multiple times.
+
+Example of **incorrect** code for this rule:
+
+```js
+  await waitFor(() => {
+    fireEvent.keyDown(input, { key: 'ArrowDown' });
+    expect(b).toEqual('b');
+  });
+
+  // or
+  await waitFor(function() {
+    fireEvent.keyDown(input, { key: 'ArrowDown' });
+    expect(b).toEqual('b');
+  });
+
+  // or
+  await waitFor(() => {
+    userEvent.keyDown(input, { key: 'ArrowDown' });
+    expect(b).toEqual('b');
+  });
+
+  // or
+  await waitFor(function() {
+    userEvent.keyDown(input, { key: 'ArrowDown' });
+    expect(b).toEqual('b');
+  });
+};
```

Comment:
a minor comment just to make the examples more realistic: there's no `keyDown` method in `userEvent` (that one belongs to `fireEvent`). 

You could replace it with some other method, such as `click`, `type` or `hover`. The same applies for the "correct" section